🐛 remove ipam pool reference defaulting to metal3

schrej opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, minor or feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:
#769 introduced defaulting for IPAM pool references that changes group and kind to metal3 IPAM. Since Metal3DataTemplates are immutable, this causes issues with existing resources when they are continuously applied (gitops).

The easiest option is to remove the defaulting in the webhook, since there are two more instances of defaulting in the DataManager ([here](https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/baremetal/metal3data_manager.go#L400) and [here](https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/baremetal/metal3data_manager.go#L434)).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1010

lentzi90 commented on api/v1beta1/metal3datatemplate_webhook.go:
```diff
@@ -39,53 +37,7 @@ func (c *Metal3DataTemplate) SetupWebhookWithManager(mgr ctrl.Manager) error {
-		}
-	}
-}
+func (c *Metal3DataTemplate) Default() {}
```
Perhaps a comment here to explain why it isn't doing anything would be good?

lentzi90 replied:
Good point. No need to add it.

lentzi90 commented:
/test-ubuntu-integration-main
/test-centos-e2e-integration-main

lentzi90 reviewed (commented):
/lgtm

mboukhalfa commented:
/test-e2e-upgrade-main-from-release-1-4

mboukhalfa commented:
/keep-test-e2e-upgrade-main-from-release-1-4

furkatgofurov7 commented:
@mboukhalfa what is blocking us to merge this, since I see required jobs are passing?

mboukhalfa commented:
> I think that is because #769 is merged into main where it introduced defaulting for IPAM pool references whereas v1.4 still uses the old defaulting.
 
 Actually #769 is implemented in 1.4.0 check here https://github.com/metal3-io/cluster-api-provider-metal3/releases/tag/v1.4.0

schrej commented:
> I think that is because #769 is merged into main where it introduced defaulting for IPAM pool references whereas v1.4 still uses the old defaulting. Also, @schrej I don't recall the reason behind, but why we had to introduce `IPAddressesFromPool` in #769 when we had `IPAddressesFromIPPool` already

I assume the question was supposed to be why we added `FromPoolRef` even though we already had `IPAddressesFromIPPool`.

The reason is that `IPAddressFromIPPool` is a string, and there is no nice way to extend that with Group/Kind information. That's why I've added a new field as an alternative.

For other fields, which were using the `FromPool` struct I just added new fields, but those fields are required, which leads to the issue we're seeing here.

mboukhalfa commented:
> For other fields, which were using the `FromPool` struct, I just added new fields, but those fields are required, which leads to the issue we're seeing here.

What confuses me is that the template applied on the bootstrap cluster works but when we apply workload template on the target cluster it fails I still do not understand the difference :/

tuminoid commented:
Overriding unrelated test failures per @mboukhalfa's request.

/override keep-test-e2e-upgrade-from-release-1-4 
/override test-e2e-upgrade-main-from-release-1-4

mboukhalfa commented:
/approve
